first changes to fix weight saving issue

ikardasevic opened the pull request:
Fixes #232 

**What was changed?**

The fix is not yet functioning but the main changes occurred in the data_collection_ui.py and data_collection_database.py files. The change_selected_value and get_values_for_date methods in data_collection_ui.py were updated to correctly interact with the DataCollectionDatabase class for saving and retrieving measurement data. The set_data_for_entry method in data_collection_database.py will be uncommented, fixed, and expanded to correctly write data into a CSV file. 

**Why was it changed?**

These changes will be necessary to fix the bug where weight measurements were not being properly saved or retrieved from the database. The original implementation is incomplete and commented out, preventing it from functioning correctly. Additionally, the DataCollectionDatabase class was not being utilized effectively in data_collection_ui.py, causing missing interactions between the UI and the data backend.

**How was it changed?**

In data_collection_ui.py, the change_selected_value and get_values_for_date methods were updated to use DataCollectionDatabase instead of ExperimentDatabase to handle the measurement data. In data_collection_database.py, the set_data_for_entry method will be uncommented, corrected, and expanded to add or update entries in a CSV file. The file creation logic was also restored to ensure a CSV is generated with appropriate headers if the file doesn’t exist. These adjustments ensure that the UI properly interacts with the database and that data is handled correctly.

moshernoah commented on databases/data_collection_database.py:
```diff
@@ -33,6 +33,15 @@ def __init__(self, experiment: str = None, measurement_items: list = ["Weight",
+        #         f.write(headers)
+
+
+
```
Is this necessary? Would this screen ever be accessed if the file is not already existing?

jackbelyeu replied:
```    #     if not path.is_file():
    #         f = open(self.filename, "x")
    #         string = "Date,Animal ID,"
    #         for item in measurement_items:
    #             string += item + ","
    #         string = string[:-1]
    #         f.write(string)
    #         f.close()```

This code is already commented on lines 25-32 and looks similar to this. If the commented code needs to be updated/deleted/uncommented you can use the pre-existing code.

moshernoah commented on databases/data_collection_database.py:
```diff
@@ -58,4 +67,23 @@ def set_data_for_entry(self, values: tuple):
+    #         new_row = pd.Series(values, index=data_file.columns)
+    #         data_file = data_file.append(new_row, ignore_index=True)
+
+    #     data_file.to_csv(self.filename, index=False)
```
Is this necessary if the export_to_csv method is already defined and functional?

jackbelyeu replied:
The original file is also all commented, so I don't think the original method ```def set_data_for_entry(self, values: tuple):``` does nothing. I agree, it shouldn't need to write anything to csv as we already have a dedicated way to export in that format. What is this method supposed to do?

moshernoah commented on experiment_pages/experiment/data_collection_ui.py:
```diff
@@ -127,6 +127,19 @@ def change_selected_value(self, values):
+        #         (str(date.today()), animal_id) + tuple(new_values)
+        #     )
+
+
```
I see how this can ensure the changes are saved, but is it needed for every increment?

jackbelyeu replied:
1. Be sure to uncomment code so the automatic workflows can check on it, and so you can validate functionality
2. Is this code intended to be a part of the method 'def change_selected_value(...)? If so, there is already ```if("None" in old_measurements):
            self.database.add_data_entry(date.today(), animal_id, new_values)
        else:
            self.database.change_data_entry(date.today(), animal_id, new_values)``` on lines 119-122, which appears to do the same thing.

moshernoah commented on experiment_pages/experiment/data_collection_ui.py:
```diff
@@ -136,6 +149,8 @@ def get_values_for_date(self, _):
 
         values = self.database.get_data_for_date(self.current_date)
+        # values = self.data_database.get_data_for_date(self.current_date)  updated code
+
```
This is fine.

ikardasevic commented:
What Was Changed?
Updated finish() Method in ChangeMeasurementsDialog. I ensured the finish() method calls get_all_values() to gather all the input values, including the weight, before passing them to self.data_collection.set_data_for_entry(). Modified get_all_values() Method to ensure it collects values from all self.textboxes entries, including the weight field, and returns them as a tuple. Updated save_temp_to_file() Method in file_utils.py to include an additional parameter, weight, with the goal of including the weight data when saving the file. The function was intended to ensure that the weight value would be somehow included in the saved data.

Why Was It Changed?
The main goal of the changes was to include weight as a data point that gets collected and saved properly when the user finishes entering data. The issue is that the weight data was not being correctly saved. I wanted to make sure that the values collected from the UI, including any empty fields, are handled correctly and formatted in a way that makes them easy to store or process later. The primary motivation behind modifying save_temp_to_file() was to ensure that the weight data collected from the UI is explicitly saved to the file. 

How Was It Changed?
I updated finish() to use the get_all_values() method to collect all values, including the weight, and then save them using self.data_collection.set_data_for_entry(values). This ensures the collected data includes the weight. In the get_all_values(), I made the method Iterate over self.textboxes to retrieve values, stripped any whitespace, and ensured that empty values default to "0". The method signature was updated to include a weight parameter: def save_temp_to_file(temp_file_path: str, permanent_file_path: str, weight: float).

moshernoah commented:
Will be closed by #296

jackbelyeu commented:
Updated work on the issue done in #296
